### DISCUSSION TITLE
Building an adjacency matrix in TF model call method

### DISCUSSION THREAD

**Discussion by relativeflux** [danielegrattarola/spektral#250] [Q&A]
I am attempting to implement the model described in [this](https://arxiv.org/pdf/2009.01599.pdf) paper on self-constructing graphs, where the graph adjacency matrix is built from the output of a variational autoencoder (more specifically the output of the encoder part). This happens in the `call` method of the decoder class, like so:

    def call(self, inputs):
        (x, z) = inputs
        a = tf.matmul(z, tf.transpose(z, perm=(0, 2, 1)))
        a = self.relu(a)
        a = spektral.utils.gcn_filter(a)
        x = self.gcn([x, a])
        # ...more code...

In the code snippet above `self.gcn` is a [`GCNConv`](https://graphneural.network/layers/convolution/#gcnconv) layer instance, which takes a modified Laplacian, computed with [`gcn_filter`](https://graphneural.network/utils/convolution/#gcn_filter). So I am attempting to run this function on z, but it's not liking it since z is a tensor, not a numpy array or sparse matrix. The problem occurs in `fit`, which runs non-eagerly, so tensors won't have a `numpy` property.

How can I solve this? Is the only way to reimplement `gcn_filter` in pure TensorFlow?

**Comment by danielegrattarola** [comment]
Hi, 

yes, `gcn_filter` is not designed to be used in a computational graph like that. 
You can re-write it from scratch, it should not be too difficult. Let me know if you need help in doing this

Cheers

**Comment by relativeflux** [comment]
@danielegrattarola Thank you Daniele. Indeed, I was looking at the code and it would seem pretty straightforward to convert it.

Fantastic library, by the way, thanks for all your hard work.
